In [1]:
#!/usr/local/bin/python3	

print ('')
print ('** Lab Big Data in Geogrpahic Information Systems **')
print ('*** Module 2: Observational data in climate sciences ***')
print ('')

## general purpose libraries
import numpy as np
#from io import StringIO
import os
import pandas as pd
#import re
#import codecs

## date libraries
import datetime as dtm
from datetime import datetime

## plotting libraries
import matplotlib.pyplot as plt
#from mpl_toolkits.basemap import Basemap



#-------------------------------------------------
#-- get info on database and fiels organization --
#-------------------------------------------------

flist = [os.path.join(path, name) for path, subdirs, files in os.walk("./CRUTEM.4.6.0.0.station_files/") for name in files]
for i in range(0,5):
  print (flist[i])
  
flist=flist[2:]
print (flist[0:5])

nst=len(flist)
print ("Number of stations = ",nst)
print ('')


#- read example obs table 

filein='./CRUTEM.4.6.0.0.station_files/61/616120'
#filein=flist[0]
print (filein)

#- scan text file to get information from the header 

with open(filein, errors='ignore') as f: data = f.readlines()
#data.encode('utf-8').strip()
print(data[3])
print (len(data))

#- find starting line of data matrix

skipi = data.index("Obs:\n")+1
print('skipi=',skipi)

#data = codecs.open(filein,'r', encoding='utf-8', errors='ignore')
#with open(filein, "r") as fp:
#  for line in fp:
#    line = bytes(line, 'utf-8').decode('utf-8', 'ignore')
#skipi = fp.index("Obs:\n")+1    


#- Extract geographical coordinates

lat_line = [line for line in data if "Lat" in line]
lat = str(lat_line).split("=",1)[1] 
lat = float(lat[:-4].strip())

lon_line = [line for line in data if "Long" in line]
lon = str(lon_line).split("=",1)[1] 
lon = float(lon[:-4].strip())

country_line = [line for line in data if "Country" in line]
country = str(country_line).split("=",1)[1] 

#print('**',lat,lon)


#- Extract data matrix

#- note that fields are separated either by a single or a double white space, 
#   which may cause troubles when trying to easily read in the data ...
#   ... using "delimiter=None" takes care of it

yr = np.genfromtxt(filein, skip_header=skipi, delimiter=None, usecols=0, dtype='i4')
print(yr)
nyr = len(yr)
print (nyr,' years of data from ',yr[0],' to ',yr[nyr-1])

mo = np.genfromtxt(filein, skip_header=skipi, filling_values=-99.0, delimiter=None, dtype='str')[:,1:13]  #[rows,cols]

print (yr)
print (mo[0:2,0:3])

#- build axis for station time series
ti0 = datetime.strptime(str(yr[0]), "%Y")
ti1 = datetime.strptime(str(yr[nyr-1]+1), "%Y")
print(ti0,ti1)
stime = pd.date_range(ti0, ti1, freq='M')
print (stime)
print (len(stime))

#- build station time series
xtemps = np.ma.empty([len(stime)],dtype='float')
cy = 0
for yi in range(0,nyr):
  for mi in range (0,12):
    xtemps[cy] = mo[yi,mi]
    cy=cy+1
    
#print (xtemps)   
xtemps[xtemps==-99.0] = np.nan
#print (xtemps)      

#plt.plot(stime,xtemps)
#plt.show()
        
#- Now build a dataframe with unique column name       
        
xdf = pd.DataFrame({'time': stime, filein[-6:]: xtemps})
xdf = xdf.set_index('time')
     
#print (xdf)        
        
#dates = datetime.strptime(dlin, "%Y-%m-%d:%H") # %HH:%MM:%SS")



#ydf = pd.read_csv(filein, skiprows=21, delim_whitespace=True,usecols=[0],header=None)
#print (ydf.dtypes)
#print(ydf.iloc[-1],ydf.iloc[-1])

#tdf = pd.read_csv(filein, skiprows=21, usecols=[1,2,3,4,5,6,7,8,9,10,11,12],header=None,dtype='float',delim_whitespace=True)
#print (tdf)
# delim_whitespace=True accounts for consecutive blanks as a unique separator


#------------------------------------------------------------
#-- Create data container with time axis from 1850 to 2018 --
#------------------------------------------------------------

#- Monthly resolution
#- Number of stations

#- Build common time axis

taxis = pd.date_range('1850-01', '2019-01', freq='M')
nmonths=len(taxis)
#print (taxis)
#print (len(taxis))

#- Build base dataframe for data

df = pd.DataFrame({'time': taxis})
df = df.set_index(['time'])


#- Build base arrays for metadata

ids = np.empty([1],dtype="str") 
olat = np.empty([1],dtype="float") 
olon = np.empty([1],dtype="float") 
ocountry = np.empty([1],dtype="str") 

#- Now loop over the stations build the database stepwise
#- For the final execution make sure to loop over the entire station list

nst_tmp = sum([len(files) for r, d, files in os.walk("/Users/davidgovi/Dropbox/Data_Science/UniMiB/Geo_Lab/CRUTEM.4.6.0.0.station_files")])

for si in range (0,200):  #range(0,nst-1):

  #- Show processing progress

  if (si==1 or si==5 or si==100 or si==1000 or si ==3000 or si ==6000 or si ==9000 or si ==nst_tmp):
    print (si)

  #- open file 
  
  filein=flist[si]
  #print (si,filein[-6:])
  
  #- scan text file to find starting line of data matrix (if data are actually present)
  with open(filein) as f: data = f.readlines()
  skipi = data.index("Obs:\n")+1
  if (len(data)-skipi < 2):
    continue
  
  #- read in data & apply some filters at the same time
  
  yr = np.genfromtxt(filein, skip_header=skipi, delimiter=None, usecols=0, dtype='i4')
  nyr = len(yr)
  if (nyr < 15):
    continue  # this filters out data series that are too short 
  if (yr[0]>1990 or yr[nyr-1]<1960):
    continue  # do not retain series not covering at least in part the reference period      
  mo = np.genfromtxt(filein, skip_header=skipi, filling_values=-99.0, delimiter=None, dtype='str')[:,1:13]  #[rows,cols]
  #print (nyr,' years of data from ',yr[0],' to ',yr[nyr-1])

  #- build local time axis

  ti0 = datetime.strptime(str(yr[0]), "%Y")
  ti1 = datetime.strptime(str(yr[nyr-1]+1), "%Y")
  stime = pd.date_range(ti0, ti1, freq='M')               

  #- build local temperature time series & clean
    
  xtemps = np.ma.empty([len(stime)],dtype='float')
  cy = 0
  for yi in range(0,nyr):
    for mi in range (0,12):
      xtemps[cy] = mo[yi,mi]
      cy=cy+1    
  xtemps[xtemps==-99.0] = np.nan
 
  #- build local station dataframe
  
  xdf = pd.DataFrame({'time': stime, filein[-6:]: xtemps})
  xdf = xdf.set_index('time')
  
  #- merge into global dataframe

  df = pd.merge(df, xdf, on='time', how='left')

  #- store metadata
  #- station name & coordinates, country
  
  ids = np.append(ids,filein[-6:])
  
  lat_line = [line for line in data if "Lat=" in line]
  lat = str(lat_line).split("=",1)[1] 
  lat = float(lat[:-4].strip())
  olat = np.append(olat,lat)

  lon_line = [line for line in data if "Long=" in line]
  lon = str(lon_line).split("=",1)[1] 
  lon = -float(lon[:-4].strip())      # all longitudes have the wrong sign... correct with -
  olon = np.append(olon,lon)

  country_line = [line for line in data if "Country" in line]
  country = str(country_line).split("=",1)[1] 
  country  = country [:-4].strip()
  ocountry = np.append(ocountry,country)

#print (df)
print(ids)
print(olat)

#- arrange metadata

#- remove leading empty element
ids = ids[1:]
olat = olat[1:]
olon = olon[1:]
ocountry = ocountry[1:]

#- put into dataframe

mdf = pd.DataFrame({'ids': ids, 'lon': olon, 'lat': olat, 'country': ocountry})
mdf = mdf.set_index('ids')

print(mdf)

#stop


#-----------------------------------------------
#-- Data treatment and visualization / Part 1 --
#-----------------------------------------------

#-- (1) Plot station locations --

import cartopy.crs as ccrs
import cartopy.feature as cfeature

ax = plt.axes(projection=ccrs.PlateCarree())
#ax = plt.axes(projection=ccrs.Robinson())

plt.title('Surface temperature measurements - global\n\n')

#- Define underlying map features

ax.set_global()
#ax.set_extent([-180,180,-90,90],crs=ccrs.Geodetic()) #(x0, x1, y0, y1) #lon here is [-180,+180] instead of [0,360]

#ax.stock_img()
ax.coastlines()

ax.add_feature(cfeature.LAND)
#ax.add_feature(cfeature.COASTLINE)
ax.add_feature(cfeature.BORDERS, edgecolor='gray')

ax.gridlines(draw_labels=True)

#- Plot data

plt.plot([olon], [olat],
         color='red', linewidth=2, marker='o', markersize=1,
         transform=ccrs.Geodetic())

# Save the plot by calling plt.savefig() BEFORE plt.show()
plt.savefig('Module2_plot01.pdf')
#plt.show()
plt.close()


#-- (2) Plot some time series of randomly chosen stations --

#- Set multipanel plot
fig = plt.figure(figsize=(9,15))

subplots = (4,2)
n_panels = subplots[0] * subplots[1]

proj = ccrs.PlateCarree()

#- Select sites
#rssite = [1,2,3,4]
rssite = np.random.randint(len(df.columns), size=4)
ssite = ids[rssite]
print(ssite)

for fi, f in enumerate(ssite):

    #- locate on map
    ax = fig.add_subplot(subplots[0], subplots[1], (fi*2)+1, projection=proj)
    ax.set_title(' / '.join([mdf.country[ssite[fi]], str(mdf.lon[ssite[fi]]) ] ), {"fontsize" : 9})
    ax.stock_img()
    ax.coastlines()
    ax.add_feature(cfeature.BORDERS)
    plt.plot(mdf.lon[ssite[fi]], mdf.lat[ssite[fi]],
         color='red', marker='o', markersize=5,
         transform=ccrs.Geodetic())         
 
    #- time series
    tser = fig.add_subplot(subplots[0], subplots[1], (fi+1)*2)
    df[ssite[fi]].plot()   

#fig.tight_layout()
plt.savefig('Module2_plot02.pdf')
#plt.show()
plt.close()


** Lab Big Data in Geogrpahic Information Systems **
*** Module 2: Observational data in climate sciences ***

./CRUTEM.4.6.0.0.station_files/.DS_Store
./CRUTEM.4.6.0.0.station_files/61/616120
./CRUTEM.4.6.0.0.station_files/61/612913
./CRUTEM.4.6.0.0.station_files/61/612770
./CRUTEM.4.6.0.0.station_files/61/614010
['./CRUTEM.4.6.0.0.station_files/61/612913', './CRUTEM.4.6.0.0.station_files/61/612770', './CRUTEM.4.6.0.0.station_files/61/614010', './CRUTEM.4.6.0.0.station_files/61/618470', './CRUTEM.4.6.0.0.station_files/61/614420']
Number of stations =  10294

./CRUTEM.4.6.0.0.station_files/61/616120
Lat=   16.6

89
skipi= 21
[1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963 1964
 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977 1978
 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992
 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006
 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018]
68  years of data fr

In [2]:
#-----------------------------------------------
#-- Data treatment and visualization / Part 2 --
#-----------------------------------------------

#-- MAKE SURE TO PROPERLY COMMENT YOUR CODE !! -- 

#-- Make a compressed file (gz or zip) with the script and the plots produced for Task 2
#-- Call it "YOURSURNAME_module2.*"
#-- Send by email (samuel.albani@unimib.it) with object "Big Data Lab in GIS - Module 2" 
#-- Upon receiving the email with attachments I will register your presence to today's lab session
#-- For those who physically attended the lab: send the email by the end of the class at 15:30
#-- For those who have specific situations (student-worker, stage abroad): contact me directly ASAP
#-- This exercise will be considered for the final evaluation, according to the criteria previously explained

#- Apply filters to data
#- Apply transformations: anomaly
#- Work on spatial component: regridding

# Task 2.1 --
# clean a bit the time series

#- only stations with complete reference period [1961-1990]
#- calculate anomaly at monthly resolution wrt reference period


# https://towardsdatascience.com/basic-time-series-manipulation-with-pandas-4432afee64ea

# task 2.1

In [3]:
#resetto l'indice in modo da avere 'time' come colonna
df = df.reset_index()

In [4]:
#definisco, e successivamente applico, una breve funzione che dal timestamp contenuto nella colonna 'time' estragga
#soltanto l'informazione relativa all'anno
def extract_year(timest):
    anno= timest.year
    return anno

In [5]:
df['year'] = df['time'].map(lambda x: extract_year(x))

In [6]:
#filtro, utilizzando l'informazione estratta in precedenza, le serie temporali mantenendo soltanto quelle comprese
#tra il 1961 e il 1990
df = df[df['year']>=1961]
df = df[df['year']<=1990]
df

time  612913  612770  614010  618470  614420  618810  619900  \
1332 1961-01-31    24.6    22.2    15.7    21.0    20.9    25.2    26.5   
1333 1961-02-28    27.6    24.3    19.6    23.0    23.9    26.9    26.8   
1334 1961-03-31    30.9    28.1    21.9    25.1    25.1    28.5    25.8   
1335 1961-04-30    32.8    32.6    24.1    25.0    28.3    27.7    24.9   
1336 1961-05-31    30.9    32.5    27.0    25.0    27.1    27.6    24.0   
1337 1961-06-30    29.1    31.0    25.8    23.7    29.5    26.0    22.8   
1338 1961-07-31    25.6    27.7    34.9    22.5    28.7    24.8    21.3   
1339 1961-08-31    24.8    26.4    34.0    22.2    29.7    24.7    20.7   
1340 1961-09-30    25.4    27.0    30.1    23.9    30.4    25.5    21.3   
1341 1961-10-31    28.9    29.5    24.9    24.1    28.4    26.1    22.5   
1342 1961-11-30    29.3    27.2    15.2    24.1    25.8    26.6    24.0   
1343 1961-12-31    26.3    23.4    20.3    22.3    23.3    26.0    25.3   
1344 1962-01-31    26.4    22.9    18.0    23.3    22.8    26.7    25.5   
1345 1962-02-28    28.3    25.9    17.4    24.0    23.1    26.6    26.1   
1346 1962-03-31    30.8    28.4    20.1    25.0    25.9    27.7    26.0   
1347 1962-04-30    31.2    31.2    23.8    24.7    25.2    27.3    24.0   
1348 1962-05-31    30.4    32.6    25.7    25.1    25.7    27.0    22.3   
1349 1962-06-30    27.4    29.7    27.5    23.4    28.1    25.6    20.3   
1350 1962-07-31    25.7    28.2    32.3    23.4    27.8    25.3    19.7   
1351 1962-08-31    24.6    26.2    25.2    22.6    29.3    24.7    20.1   
1352 1962-09-30    25.7    26.5    30.9    23.5    29.5    25.6    21.2   
1353 1962-10-31    27.8    28.5    25.3    24.3    29.9    26.2    21.9   
1354 1962-11-30    28.0    27.8    19.1    24.4    27.6    26.4    22.9   
1355 1962-12-31    24.8    23.0    17.3    22.7    23.0    25.7    24.7   
1356 1963-01-31    28.2    25.3    17.9    23.6    23.5    26.4    25.3   
1357 1963-02-28    28.5    29.0    18.3    24.8    23.6    26.7    25.5   
1358 1963-03-31    29.9    28.4    22.0    24.7    26.3    27.7    25.0   
1359 1963-04-30    31.9    32.0    22.6    25.3    24.8    27.7    24.3   
1360 1963-05-31    30.7    31.7    23.2    25.4    28.1    27.1    22.1   
1361 1963-06-30    30.0    31.5    25.7    24.7    28.3    26.3    21.8   
...         ...     ...     ...     ...     ...     ...     ...     ...   
1662 1988-07-31     NaN    27.9    32.5    23.4    26.7    25.0    21.6   
1663 1988-08-31     NaN    27.1    34.2    24.0     NaN    24.9    21.3   
1664 1988-09-30     NaN    27.6    30.5    24.1    28.7    25.7    21.9   
1665 1988-10-31     NaN    29.8     NaN    24.5    27.8    27.4    23.3   
1666 1988-11-30     NaN    27.6    21.0    24.6    24.8    27.2    24.5   
1667 1988-12-31     NaN    23.2    15.5    22.1    21.3     NaN    25.2   
1668 1989-01-31     NaN    23.6    14.9    21.3    21.4     NaN    26.5   
1669 1989-02-28     NaN    25.9    18.3    23.6     NaN     NaN    26.2   
1670 1989-03-31     NaN    30.0    22.0    25.5    23.1     NaN    25.9   
1671 1989-04-30     NaN    33.3    20.7    25.9    23.7     NaN    24.6   
1672 1989-05-31     NaN    34.1    25.3    25.7    25.9     NaN    23.3   
1673 1989-06-30     NaN    30.9    26.9    24.3    28.8     NaN    21.9   
1674 1989-07-31     NaN    27.9    32.4    23.8     NaN     NaN    21.3   
1675 1989-08-31     NaN    26.7     NaN     NaN     NaN     NaN    20.8   
1676 1989-09-30     NaN    28.1    30.9    24.0    28.9     NaN    21.3   
1677 1989-10-31     NaN    28.9    27.4    24.4    29.6     NaN    22.6   
1678 1989-11-30     NaN    28.6    20.9    24.1    24.1     NaN    23.4   
1679 1989-12-31     NaN    25.7     NaN    22.3    23.5     NaN    25.1   
1680 1990-01-31     NaN    24.2     NaN    22.5    20.2     NaN    25.2   
1681 1990-02-28     NaN    26.9     NaN    23.5    25.1     NaN    25.6   
1682 1990-03-31     NaN    29.9     NaN    24.0    27.7     NaN    25.5   
1683 1990-04-30     NaN    34.4     

In [7]:
#elimino tutte le le colonne relative alle stazioni meteorologiche contenenti valori NaN
df = df.dropna(axis='columns')
df

time  612770  612650  612260  612850  612930  612330  612300  \
1332 1961-01-31    22.2    21.7    22.6    25.6    21.6    22.1    20.7   
1333 1961-02-28    24.3    24.0    24.3    28.6    23.1    24.6    22.9   
1334 1961-03-31    28.1    27.5    28.1    31.2    26.9    27.9    27.7   
1335 1961-04-30    32.6    31.4    30.7    32.5    31.3    30.2    30.8   
1336 1961-05-31    32.5    33.6    36.7    33.6    31.0    34.4    34.9   
1337 1961-06-30    31.0    31.7    35.3    29.8    29.2    32.7    32.9   
1338 1961-07-31    27.7    28.4    32.6    27.0    25.9    29.3    28.5   
1339 1961-08-31    26.4    27.4    30.6    26.3    25.4    27.9    27.3   
1340 1961-09-30    27.0    28.2    32.7    26.8    26.1    29.1    28.7   
1341 1961-10-31    29.5    28.8    32.8    27.9    27.3    30.7    31.1   
1342 1961-11-30    27.2    26.5    28.7    26.9    24.2    27.8    27.4   
1343 1961-12-31    23.4    22.5    23.3    24.8    20.8    23.4    22.4   
1344 1962-01-31    22.9    22.5    22.8    26.3    21.0    22.5    22.7   
1345 1962-02-28    25.9    26.0    25.9    29.7    24.6    25.3    24.8   
1346 1962-03-31    28.4    27.6    28.8    31.2    26.9    26.9    27.1   
1347 1962-04-30    31.2    30.7    31.5    32.6    29.2    30.4    31.4   
1348 1962-05-31    32.6    33.1    36.8    31.7    30.3    34.0    34.4   
1349 1962-06-30    29.7    31.4    35.6    28.6    28.1    31.9    31.9   
1350 1962-07-31    28.2    28.8    33.6    26.9    26.9    29.3    28.9   
1351 1962-08-31    26.2    26.4    30.6    26.3    25.7    27.4    26.9   
1352 1962-09-30    26.5    27.9    32.9    26.6    25.6    28.8    28.4   
1353 1962-10-31    28.5    29.3    34.1    27.8    27.3    29.5    29.5   
1354 1962-11-30    27.8    27.1    30.6    27.4    25.8    27.7    27.0   
1355 1962-12-31    23.0    21.7    23.9    25.2    21.7    21.9    21.2   
1356 1963-01-31    25.3    24.5    25.6    27.3    22.7    24.1    24.5   
1357 1963-02-28    29.0    28.5    30.4    29.5    26.9    27.3    26.6   
1358 1963-03-31    28.4    27.7    29.1    30.6    27.5    26.2    26.8   
1359 1963-04-30    32.0    31.1    31.9    32.3    30.5    30.1    31.0   
1360 1963-05-31    31.7    30.9    34.5    32.6    30.2    30.8    31.9   
1361 1963-06-30    31.5    32.0    35.1    31.1    29.9    32.9    33.0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
1662 1988-07-31    27.9    29.6    32.7    25.2    26.0    29.4    29.8   
1663 1988-08-31    27.1    28.4    31.7    25.2    25.4    27.7    27.3   
1664 1988-09-30    27.6    28.9    32.3    25.6    26.5    28.8    28.1   
1665 1988-10-31    29.8    31.0    32.1    26.6    28.5    30.0    30.4   
1666 1988-11-30    27.6    27.6    27.3    25.0    26.3    25.7    26.2   
1667 1988-12-31    23.2    22.7    22.1    23.4    23.0    21.1    21.3   
1668 1989-01-31    23.6    23.3    22.1    25.1    23.1    23.0    22.5   
1669 1989-02-28    25.9    25.9    24.0    28.1    25.4    25.3    25.7   
1670 1989-03-31    30.0    29.7    28.9    29.4    29.1    28.1    27.8   
1671 1989-04-30    33.3    33.2    32.8    31.4    32.3    31.6    30.6   
1672 1989-05-31    34.1    34.8    35.8    32.0    32.2    34.2    34.7   
1673 1989-06-30    30.9    32.8    36.4    28.5    29.2    33.7    33.6   
1674 1989-07-31    27.9    29.5    32.8    26.0    26.5    28.4    28.9   
1675 1989-08-31    26.7    27.4    30.7    25.8    25.6    27.0    27.4   
1676 1989-09-30    28.1    30.2    33.5    26.7    26.8    30.0    29.7   
1677 1989-10-31    28.9    30.7    32.4    27.3    28.0    30.1    30.0   
1678 1989-11-30    28.6    29.9    28.7    27.3    27.3    28.8    28.9   
1679 1989-12-31    25.7    26.1    24.8    25.1    24.7    24.3    24.9   
1680 1990-01-31    24.2    23.9    23.7    25.0    24.2    21.9    21.3   
1681 1990-02-28    26.9    27.0    25.2    28.7    26.4    26.1    26.2   
1682 1990-03-31    29.9    30.4    29.1    31.2    29.3    29.8    30.9   
1683 1990-04-30    34.4    34.4    3

In [10]:
#droppo le colonne 'time' e 'year' in modo che non interferiscano col calcolo.
#per ogni stazione calcolo la media e la sottraggo al valore del periodo preso in osservazione
df.iloc[:,1:-1].apply(lambda x:x-np.mean(x))

612770    612650    612260    612850    612930  612330    612300  \
1332 -6.216111 -7.032222 -7.620278 -2.918056 -5.535833  -6.515 -7.849444   
1333 -4.116111 -4.732222 -5.920278  0.081944 -4.035833  -4.015 -5.649444   
1334 -0.316111 -1.232222 -2.120278  2.681944 -0.235833  -0.715 -0.849444   
1335  4.183889  2.667778  0.479722  3.981944  4.164167   1.585  2.250556   
1336  4.083889  4.867778  6.479722  5.081944  3.864167   5.785  6.350556   
1337  2.583889  2.967778  5.079722  1.281944  2.064167   4.085  4.350556   
1338 -0.716111 -0.332222  2.379722 -1.518056 -1.235833   0.685 -0.049444   
1339 -2.016111 -1.332222  0.379722 -2.218056 -1.735833  -0.715 -1.249444   
1340 -1.416111 -0.532222  2.479722 -1.718056 -1.035833   0.485  0.150556   
1341  1.083889  0.067778  2.579722 -0.618056  0.164167   2.085  2.550556   
1342 -1.216111 -2.232222 -1.520278 -1.618056 -2.935833  -0.815 -1.149444   
1343 -5.016111 -6.232222 -6.920278 -3.718056 -6.335833  -5.215 -6.149444   
1344 -5.516111 -6.232222 -7.420278 -2.218056 -6.135833  -6.115 -5.849444   
1345 -2.516111 -2.732222 -4.320278  1.181944 -2.535833  -3.315 -3.749444   
1346 -0.016111 -1.132222 -1.420278  2.681944 -0.235833  -1.715 -1.449444   
1347  2.783889  1.967778  1.279722  4.081944  2.064167   1.785  2.850556   
1348  4.183889  4.367778  6.579722  3.181944  3.164167   5.385  5.850556   
1349  1.283889  2.667778  5.379722  0.081944  0.964167   3.285  3.350556   
1350 -0.216111  0.067778  3.379722 -1.618056 -0.235833   0.685  0.350556   
1351 -2.216111 -2.332222  0.379722 -2.218056 -1.435833  -1.215 -1.649444   
1352 -1.916111 -0.832222  2.679722 -1.918056 -1.535833   0.185 -0.149444   
1353  0.083889  0.567778  3.879722 -0.718056  0.164167   0.885  0.950556   
1354 -0.616111 -1.632222  0.379722 -1.118056 -1.335833  -0.915 -1.549444   
1355 -5.416111 -7.032222 -6.320278 -3.318056 -5.435833  -6.715 -7.349444   
1356 -3.116111 -4.232222 -4.620278 -1.218056 -4.435833  -4.515 -4.049444   
1357  0.583889 -0.232222  0.179722  0.981944 -0.235833  -1.315 -1.949444   
1358 -0.016111 -1.032222 -1.120278  2.081944  0.364167  -2.415 -1.749444   
1359  3.583889  2.367778  1.679722  3.781944  3.364167   1.485  2.450556   
1360  3.283889  2.167778  4.279722  4.081944  3.064167   2.185  3.350556   
1361  3.083889  3.267778  4.879722  2.581944  2.764167   4.285  4.450556   
...        ...       ...       ...       ...       ...     ...       ...   
1662 -0.516111  0.867778  2.479722 -3.318056 -1.135833   0.785  1.250556   
1663 -1.316111 -0.332222  1.479722 -3.318056 -1.735833  -0.915 -1.249444   
1664 -0.816111  0.167778  2.079722 -2.918056 -0.635833   0.185 -0.449444   
1665  1.383889  2.267778  1.879722 -1.918056  1.364167   1.385  1.850556   
1666 -0.816111 -1.132222 -2.920278 -3.518056 -0.835833  -2.915 -2.349444   
1667 -5.216111 -6.032222 -8.120278 -5.118056 -4.135833  -7.515 -7.249444   
1668 -4.816111 -5.432222 -8.120278 -3.418056 -4.035833  -5.615 -6.049444   
1669 -2.516111 -2.832222 -6.220278 -0.418056 -1.735833  -3.315 -2.849444   
1670  1.583889  0.967778 -1.320278  0.881944  1.964167  -0.515 -0.749444   
1671  4.883889  4.467778  2.579722  2.881944  5.164167   2.985  2.050556   
1672  5.683889  6.067778  5.579722  3.481944  5.064167   5.585  6.150556   
1673  2.483889  4.067778  6.179722 -0.018056  2.064167   5.085  5.050556   
1674 -0.516111  0.767778  2.579722 -2.518056 -0.635833  -0.215  0.350556   
1675 -1.716111 -1.332222  0.479722 -2.718056 -1.535833  -1.615 -1.149444   
1676 -0.316111  1.467778  3.279722 -1.818056 -0.335833   1.385  1.150556   
1677  0.483889  1.967778  2.179722 -1.218056  0.864167   1.485  1.450556   
1678  0.183889  1.167778 -1.520278 -1.218056  0.164167   0.185  0.350556   
1679 -2.716111 -2.632222 -5.420278 -3.418056 -2.435833  -4.315 -3.649444   
1680 -4.216111 -4.832222 -6.520278 -3.518056 -2.935833  -6.715 -7.249444   
1681 -1.516111 -1.732222 -5.020278  0.181944 -0.735833  -2.515 -2.349444   
1682  1.483889  1.667778 -1.120278  2.681944  2.164167  

# task 2.2

Il task 2.2 non è stato eseguito entro la fine del laboratorio.
Il ragionamento che avrei fatto sarebbe stato di arrotondare i valori di latitudine e longitudine delle varie stazioni in modo da ricondurle a punti comuni aggregati.
Dopo di che avrei raggruppato le serie temporali sull'anno, recuperando l'informazione utilizzata al task 2.1
Infine avrei calcolato la media globale di tutte le stazioni aggregate in precedenza, anno per anno in maniera da plottare l'andamento nel tempo. Avrei affiancato questo plot con una mappa cloropletica per mostrare quali aree concorrono, e in che modo, alla media mostrata.